File: Front/server/api/video-proxy.ts
export default defineEventHandler(async (event) => {
  // Handle OPTIONS request for CORS preflight
  if (event.node.req.method === 'OPTIONS') {
    setHeader(event, 'Access-Control-Allow-Origin', '*')
    setHeader(event, 'Access-Control-Allow-Methods', 'GET, HEAD, OPTIONS')
    setHeader(event, 'Access-Control-Allow-Headers', 'Range, Content-Type')
    setHeader(event, 'Access-Control-Max-Age', '86400')
    return ''
  }

  const query = getQuery(event)
  const videoUrl = query.url as string

  if (!videoUrl) {
    throw createError({
      statusCode: 400,
      statusMessage: 'Video URL is required'
    })
  }

  // Validate URL
  try {
    new URL(videoUrl)
  } catch {
    throw createError({
      statusCode: 400,
      statusMessage: 'Invalid video URL'
    })
  }

  try {
    // دریافت Range header از request
    const rangeHeader = getHeader(event, 'range') || ''
    
    // دریافت ویدیو از URL اصلی
    const fetchOptions: RequestInit = {
      method: 'GET',
      headers: {} as HeadersInit
    }

    // اگر Range header وجود دارد، آن را ارسال کن
    if (rangeHeader) {
      fetchOptions.headers = {
        'Range': rangeHeader
      }
    }

    const response = await fetch(videoUrl, fetchOptions)

    if (!response.ok && response.status !== 206) {
      throw createError({
        statusCode: response.status,
        statusMessage: 'Failed to fetch video'
      })
    }

    // تنظیم CORS headers
    setHeader(event, 'Access-Control-Allow-Origin', '*')
    setHeader(event, 'Access-Control-Allow-Methods', 'GET, HEAD, OPTIONS')
    setHeader(event, 'Access-Control-Allow-Headers', 'Range, Content-Type')
    setHeader(event, 'Access-Control-Expose-Headers', 'Content-Range, Content-Length, Accept-Ranges')
    setHeader(event, 'Accept-Ranges', 'bytes')

    // کپی کردن headers از response اصلی
    const contentType = response.headers.get('content-type')
    if (contentType) {
      setHeader(event, 'Content-Type', contentType)
    } else {
      // اگر Content-Type وجود ندارد، بر اساس extension حدس بزن
      const urlObj = new URL(videoUrl)
      const pathname = urlObj.pathname.toLowerCase()
      if (pathname.endsWith('.mp4')) {
        setHeader(event, 'Content-Type', 'video/mp4')
      } else if (pathname.endsWith('.webm')) {
        setHeader(event, 'Content-Type', 'video/webm')
      } else if (pathname.endsWith('.ogg')) {
        setHeader(event, 'Content-Type', 'video/ogg')
      }
    }

    const contentLength = response.headers.get('content-length')
    if (contentLength) {
      setHeader(event, 'Content-Length', contentLength)
    }

    const contentRange = response.headers.get('content-range')
    if (contentRange) {
      setHeader(event, 'Content-Range', contentRange)
      setResponseStatus(event, 206) // Partial Content
    } else if (response.status === 206) {
      setResponseStatus(event, 206)
    }

    // برای streaming، response body را مستقیماً برگردانیم
    // H3/Nitro به صورت خودکار ReadableStream را handle می‌کند
    if (response.body) {
      // تبدیل Web ReadableStream به Node.js Readable
      const { Readable } = await import('stream')
      const nodeStream = Readable.fromWeb(response.body as any)
      return nodeStream
    }
    
    // Fallback: اگر body وجود ندارد
    throw createError({
      statusCode: 500,
      statusMessage: 'Video stream not available'
    })
  } catch (error: any) {
    console.error('Error proxying video:', error)
    throw createError({
      statusCode: error.statusCode || 500,
      statusMessage: error.message || 'Error proxying video'
    })
  }
})

File: backend/core/video_proxy.py
"""
Video Proxy View برای proxy کردن ویدیوهای خارجی (مثل Aparat)
با پشتیبانی از CORS headers و Range requests
"""
import requests
import re
from django.http import HttpResponse, StreamingHttpResponse, Http404
from django.conf import settings
from rest_framework.views import APIView
from rest_framework.response import Response
from rest_framework import status


class VideoProxyView(APIView):
    """
    Proxy view برای ویدیوهای خارجی با CORS headers
    """
    
    def get(self, request):
        """
        Proxy کردن ویدیو از URL خارجی
        Query parameter: url - URL ویدیو برای proxy کردن
        """
        video_url = request.GET.get('url')
        
        if not video_url:
            return Response(
                {'error': 'URL parameter is required'},
                status=status.HTTP_400_BAD_REQUEST
            )
        
        try:
            # دریافت CORS origin از settings
            cors_origin = getattr(settings, 'VIDEO_CORS_ORIGIN', '*')
            
            # دریافت Range header از request
            range_header = request.META.get('HTTP_RANGE', '').strip()
            
            # تنظیم headers برای request به سرور خارجی
            headers = {
                'User-Agent': request.META.get('HTTP_USER_AGENT', 'Mozilla/5.0'),
                'Referer': video_url,
            }
            
            if range_header:
                headers['Range'] = range_header
            
            # درخواست به سرور خارجی
            response = requests.get(
                video_url,
                headers=headers,
                stream=True,
                timeout=30
            )
            
            # بررسی status code
            if response.status_code not in [200, 206]:
                return Response(
                    {'error': f'Failed to fetch video: {response.status_code}'},
                    status=status.HTTP_502_BAD_GATEWAY
                )
            
            # تنظیم CORS headers
            cors_headers = {
                'Access-Control-Allow-Origin': cors_origin,
                'Access-Control-Allow-Methods': 'GET, HEAD, OPTIONS',
                'Access-Control-Allow-Headers': 'Range',
                'Access-Control-Expose-Headers': 'Content-Range, Content-Length, Accept-Ranges',
                'Accept-Ranges': 'bytes',
            }
            
            # کپی کردن headers مهم از response اصلی
            content_type = response.headers.get('Content-Type', 'video/mp4')
            content_length = response.headers.get('Content-Length')
            content_range = response.headers.get('Content-Range')
            accept_ranges = response.headers.get('Accept-Ranges', 'bytes')
            
            # اگر Range request بود و 206 Partial Content برگشت
            if response.status_code == 206:
                # استفاده از StreamingHttpResponse برای partial content
                django_response = StreamingHttpResponse(
                    response.iter_content(chunk_size=8192),
                    status=206
                )
                
                # تنظیم headers
                django_response['Content-Type'] = content_type
                if content_length:
                    django_response['Content-Length'] = content_length
                if content_range:
                    django_response['Content-Range'] = content_range
                django_response['Accept-Ranges'] = accept_ranges
                
                # اضافه کردن CORS headers
                for key, value in cors_headers.items():
                    django_response[key] = value
                
                return django_response
            else:
                # برای full content
                django_response = StreamingHttpResponse(
                    response.iter_content(chunk_size=8192),
                    status=200
                )
                
                django_response['Content-Type'] = content_type
                if content_length:
                    django_response['Content-Length'] = content_length
                django_response['Accept-Ranges'] = accept_ranges
                
                # اضافه کردن CORS headers
                for key, value in cors_headers.items():
                    django_response[key] = value
                
                return django_response
                
        except requests.exceptions.RequestException as e:
            return Response(
                {'error': f'Failed to proxy video: {str(e)}'},
                status=status.HTTP_502_BAD_GATEWAY
            )
        except Exception as e:
            return Response(
                {'error': f'Internal server error: {str(e)}'},
                status=status.HTTP_500_INTERNAL_SERVER_ERROR
            )
    
    def options(self, request):
        """
        Handle OPTIONS request (CORS preflight)
        """
        cors_origin = getattr(settings, 'VIDEO_CORS_ORIGIN', '*')
        response = HttpResponse(status=200)
        response['Access-Control-Allow-Origin'] = cors_origin
        response['Access-Control-Allow-Methods'] = 'GET, HEAD, OPTIONS'
        response['Access-Control-Allow-Headers'] = 'Range'
        response['Access-Control-Expose-Headers'] = 'Content-Range, Content-Length, Accept-Ranges'
        return response
    
    def head(self, request):
        """
        Handle HEAD request
        """
        video_url = request.GET.get('url')
        if not video_url:
            return HttpResponse(status=400)
        
        try:
            response = requests.head(video_url, timeout=10)
            cors_origin = getattr(settings, 'VIDEO_CORS_ORIGIN', '*')
            
            django_response = HttpResponse(status=response.status_code)
            django_response['Access-Control-Allow-Origin'] = cors_origin
            django_response['Access-Control-Allow-Methods'] = 'GET, HEAD, OPTIONS'
            django_response['Access-Control-Allow-Headers'] = 'Range'
            django_response['Access-Control-Expose-Headers'] = 'Content-Range, Content-Length, Accept-Ranges'
            
            if 'Content-Length' in response.headers:
                django_response['Content-Length'] = response.headers['Content-Length']
            if 'Content-Type' in response.headers:
                django_response['Content-Type'] = response.headers['Content-Type']
            if 'Accept-Ranges' in response.headers:
                django_response['Accept-Ranges'] = response.headers['Accept-Ranges']
            
            return django_response
        except Exception as e:
            return HttpResponse(status=502)

